File: src/algorithms/sorting/selection/double-selection-sort/double-selection-sort.test.ts
import { describe, it, expect } from "vitest";
import { doubleSelectionSort } from "./sources/double-selection-sort.ts?fn";

describe("doubleSelectionSort", () => {
  it("sorts an unsorted array", () => {
    expect(doubleSelectionSort([64, 34, 25, 12, 22, 11, 90])).toEqual([11, 12, 22, 25, 34, 64, 90]);
  });

  it("handles an already sorted array", () => {
    expect(doubleSelectionSort([1, 2, 3, 4, 5])).toEqual([1, 2, 3, 4, 5]);
  });

  it("handles a reverse-sorted array", () => {
    expect(doubleSelectionSort([5, 4, 3, 2, 1])).toEqual([1, 2, 3, 4, 5]);
  });

  it("handles an array with duplicate values", () => {
    expect(doubleSelectionSort([3, 1, 4, 1, 5, 9, 2, 6, 5])).toEqual([1, 1, 2, 3, 4, 5, 5, 6, 9]);
  });

  it("handles a single element array", () => {
    expect(doubleSelectionSort([42])).toEqual([42]);
  });

  it("handles an empty array", () => {
    expect(doubleSelectionSort([])).toEqual([]);
  });

  it("handles an array with negative numbers", () => {
    expect(doubleSelectionSort([3, -1, 0, -5, 2])).toEqual([-5, -1, 0, 2, 3]);
  });

  it("handles an even-length array", () => {
    expect(doubleSelectionSort([4, 2, 6, 1])).toEqual([1, 2, 4, 6]);
  });

  it("does not mutate the original array", () => {
    const original = [3, 1, 2];
    const sorted = doubleSelectionSort(original);
    expect(sorted).toEqual([1, 2, 3]);
    expect(original).toEqual([3, 1, 2]);
  });
});

File: src/algorithms/sorting/selection/double-selection-sort/sources/DoubleSelectionSort_test.cpp
#include "DoubleSelectionSort.cpp"
#include <cassert>
#include <vector>
#include <iostream>

int main() {
    // sorts an unsorted array
    assert((doubleSelectionSort({64, 34, 25, 12, 22, 11, 90}) == std::vector<int>{11, 12, 22, 25, 34, 64, 90}));

    // handles an already sorted array
    assert((doubleSelectionSort({1, 2, 3, 4, 5}) == std::vector<int>{1, 2, 3, 4, 5}));

    // handles a reverse-sorted array
    assert((doubleSelectionSort({5, 4, 3, 2, 1}) == std::vector<int>{1, 2, 3, 4, 5}));

    // handles an array with duplicate values
    assert((doubleSelectionSort({3, 1, 4, 1, 5, 9, 2, 6, 5}) == std::vector<int>{1, 1, 2, 3, 4, 5, 5, 6, 9}));

    // handles a single element array
    assert((doubleSelectionSort({42}) == std::vector<int>{42}));

    // handles an empty array
    assert((doubleSelectionSort({}) == std::vector<int>{}));

    // handles an array with negative numbers
    assert((doubleSelectionSort({3, -1, 0, -5, 2}) == std::vector<int>{-5, -1, 0, 2, 3}));

    // handles an even-length array
    assert((doubleSelectionSort({4, 2, 6, 1}) == std::vector<int>{1, 2, 4, 6}));

    // does not mutate the original array
    std::vector<int> original = {3, 1, 2};
    std::vector<int> sorted = doubleSelectionSort(original);
    assert((sorted == std::vector<int>{1, 2, 3}));
    assert((original == std::vector<int>{3, 1, 2}));

    std::cout << "All tests passed!" << std::endl;
    return 0;
}
